### ISSUE TITLE
ESI nonce hole-punching breaks Gravity Forms AJAX submissions when nonce lifetime is customized

### ISSUE THREAD

**Issue by robertstaddon** [litespeedtech/lscache_wp#1005]
## Summary

LiteSpeed Cache's ESI nonce hole-punching appears to generate nonce values using WordPress's default nonce lifetime, but some plugins customize nonce lifetime per action through WordPress's `nonce_life` filter.

This causes ESI-generated nonces to fail server-side verification when the target plugin verifies the nonce using a different, action-specific lifetime.

We reproduced this with Gravity Forms AJAX submissions using the Stripe card field. Gravity Forms changes the lifetime of the `gform_ajax_submission` nonce to 3 days. LiteSpeed ESI serves a nonce value that appears to be generated on the default WordPress 1-day nonce tick. When the form submits to `admin-ajax.php`, Gravity Forms verifies that same nonce using its 3-day lifetime, so the nonce value does not match and the submission fails with:

> Your session has expired. Please refresh the page and try again.

## Environment

- WordPress
- LiteSpeed Cache with ESI enabled
- Gravity Forms
- Gravity Forms Stripe Add-On
- A Gravity Forms payment form using the Stripe card field / AJAX submission path

The issue was reproduced with Cloudflare bypassed and with LiteSpeed page caching disabled for the page. The issue persisted as long as ESI remained enabled. Disabling ESI immediately restored the correct nonce behavior.

## Relevant WordPress behavior

WordPress nonce values are derived from a time-based tick:

```php
$tick = ceil( time() / ( $nonce_life / 2 ) );
```

The nonce lifetime is filterable:

```php
apply_filters( 'nonce_life', DAY_IN_SECONDS, $action );
```

This means two nonce actions can legitimately have different lifetimes and therefore different ticks at the same point in time.

## Relevant Gravity Forms behavior

Gravity Forms customizes the nonce lifetime for AJAX submission and validation:

```php
public function nonce_life( $lifespan_in_seconds, $action = '' ) {
	if ( in_array( $action, array( 'gform_ajax_submission', 'gform_ajax_validation' ) ) ) {
		$lifespan_in_seconds = apply_filters( 'gform_nonce_life', 3 * DAY_IN_SECONDS, $action );
	}

	return $lifespan_in_seconds;
}
```

So `gform_ajax_submission` uses a 3-day lifetime instead of WordPress's default 1-day lifetime.

That changes the nonce tick from a 12-hour window to a 36-hour window, which produces a different nonce value.

## Reproduction

1. Enable LiteSpeed Cache ESI.
2. Add `gform_ajax_submission private` to the LiteSpeed ESI nonce list.
3. Load a Gravity Forms page using AJAX submission with a Stripe card field.
4. Submit the form.

Expected result:

- The ESI-generated `gform_ajax_submission` nonce verifies successfully.
- The form submits normally.

Actual result:

- The form POST includes `gform_ajax_nonce`.
- The value matches the `ajax_submission_nonce` embedded in the page HTML.
- Gravity Forms rejects it server-side.
- The submission fails with:

> Your session has expired. Please refresh the page and try again.

## Evidence observed

With ESI enabled, the page served:

```text
ajax_submission_nonce: 16a9748de8
```

The browser submitted the same value:

```text
gform_ajax_nonce: 16a9748de8
```

So the nonce was not missing from the payload and was not being mangled by JavaScript.

However, Gravity Forms rejected it during `check_ajax_referer( 'gform_ajax_submission', 'gform_ajax_nonce' )`.

When ESI was disabled, the same page served a different nonce:

```text
ajax_submission_nonce: 13823dbc59
```

That value submitted successfully.

This points to a nonce generation mismatch:

- ESI-generated value: generated using the default 1-day nonce lifetime.
- Gravity Forms verification value: generated using Gravity Forms' custom 3-day nonce lifetime.

We also confirmed that excluding the page from LiteSpeed page cache did not resolve the issue. The page-cache exclusion prevented page caching, but ESI nonce substitution still ran on the response. Only disabling ESI changed the nonce back to the value Gravity Forms expected.

## Why page-cache TTL is not the right fix

This is not primarily a page-cache TTL problem. It is a nonce generation/verification mismatch.

Even if the public cache TTL is changed, the ESI nonce generator still needs to generate the nonce using the same effective `nonce_life` that WordPress will use when verifying that nonce action.

ESI should allow a page to remain cached for a long TTL while replacing small dynamic fragments such as nonces at request time. That works only if the ESI nonce replacement honors the target action's actual lifetime.

## Suggested LiteSpeed Cache core fix

LiteSpeed Cache's ESI nonce generation should be action-aware and lifetime-aware.

When generating an ESI nonce for a configured action, LiteSpeed should generate it in a normal WordPress context where plugin filters are available and should respect the effective nonce lifetime for that specific action.

Conceptually:

```php
$life = apply_filters( 'nonce_life', DAY_IN_SECONDS, $action );
$nonce = wp_create_nonce( $action );
```

The important part is that the ESI-generated nonce and the later `check_ajax_referer()` verification must both use the same effective nonce lifetime and therefore the same nonce tick.

The ESI nonce fragment cache key and TTL should also account for:

- nonce action
- user/session context
- effective nonce lifetime
- current nonce tick

The ESI nonce fragment should not be cached longer than the tick window for the effective lifetime of that nonce action.

## Operational workaround

As a site-level workaround, we normalized Gravity Forms' AJAX nonce lifetime back to WordPress's default 1-day lifetime:

```php
add_filter( 'gform_nonce_life', function () {
	return DAY_IN_SECONDS;
} );
```

This makes Gravity Forms' `gform_ajax_submission` and `gform_ajax_validation` nonces use the same lifetime LiteSpeed ESI appears to assume, so the ESI-generated nonce verifies successfully.

This workaround is acceptable for payment forms because a user is expected to complete the payment in the same session. A form left open for more than about a day may require a refresh, which is consistent with WordPress's default nonce behavior.

However, this is only a workaround. Ideally, LiteSpeed Cache's ESI nonce implementation should support valid WordPress nonce actions whose lifetime is customized by plugins.

## Why this matters

Plugins are allowed to customize nonce lifetime using WordPress's `nonce_life` filter. Gravity Forms is one example, but other plugins may do the same.

If LiteSpeed ESI assumes all nonces use the default lifetime, ESI nonce hole-punching can silently generate invalid nonce values for legitimate plugin nonce actions.

The desired behavior is that any nonce generated through LiteSpeed ESI should verify exactly as if that nonce had been generated inline by `wp_create_nonce( $action )` during the original page render.

**Comment by robertstaddon** [comment]
Here's a pull request to resolve this issue: https://github.com/litespeedtech/lscache_wp/pull/1006

**Comment by timotei-litespeed** [comment]
@robertstaddon we are looking into the changes you added. Will be back with an answer

Thank you!
